app/lib/shared/storage.dart

```diff
@@ -0,0 +1,199 @@
+  try {
+    await bucket.delete(objectName);
+  } on DetailedApiRequestError catch (e) {
+    if (e.status != 404) {
+      rethrow;
+    }
+  }
+}
+
+/// Uploads content from [openStream] to the [bucket] as [objectName].
+Future uploadWithRetry(Bucket bucket, String objectName, int length,
+    Stream<List<int>> openStream()) async {
+  await retryAsync(
+    () async {
+      final sink = bucket.write(objectName,
+          length: length, contentType: contentType(objectName));
+      await sink.addStream(openStream());
+      await sink.close();
+    },
+    description: 'Upload to $objectName',
+    shouldRetryOnError: (e) {
+      if (e is DetailedApiRequestError) {
+        return e.status == 502 || e.status == 503;
+      }
+      return false;
+    },
+    sleep: Duration(seconds: 10),
+  );
+}
+
+/// Uploads content from [bytes] to the [bucket] as [objectName].
+Future uploadBytesWithRetry(
+        Bucket bucket, String objectName, List<int> bytes) =>
+    uploadWithRetry(bucket, objectName, bytes.length,
+        () => new Stream.fromIterable([bytes]));
+
+/// Utility class to access versioned data file that follows the pattern:
+/// "/path-prefix/runtime-version.ext". The extension must use .gz as the
+/// storage uses transparent gzip on the data stream.
+class VersionedDataStorage {
+  final Bucket _bucket;
+  final String _prefix;
+  final String _extension;
+
+  VersionedDataStorage(Bucket bucket, String prefix, String extension)
+      : _bucket = bucket,
+        _prefix = prefix,
+        _extension = extension {
+    assert(extension.endsWith('.gz'));
+  }
+
+  /// Whether the storage bucket has a data file for the current runtime version.
+  /// TODO: decide whether we should re-generate the file after a certain age
+  Future<bool> hasCurrentData() async {
+    try {
+      final info = await _bucket.info(_objectName());
+      return info != null;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Upload the current data file to the storage bucket.
+  Future uploadDataFile(File file) async {
+    final objectName = _objectName();
+    try {
+      // Prevents edge cases like double compression.
+      assert(!file.path.endsWith('.gz'));
+      await uploadWithRetry(_bucket, objectName, file.lengthSync(),
+          () => file.openRead().transform(_gzip.encoder));
+    } catch (e, st) {
+      _logger.warning('Unable to upload data file: $objectName', e, st);
+    }
+  }
+
+  /// Upload the current data to the storage bucket.
+  Future uploadDataAsJsonMap(Map<String, dynamic> map) async {
+    final objectName = _objectName();
+    final bytes = utf8.encode(json.encode(map));
+    try {
+      await uploadBytesWithRetry(_bucket, objectName, bytes);
+    } catch (e, st) {
+      _logger.warning('Unable to upload data file: $objectName', e, st);
+    }
+  }
+
+  /// Gets the content of the data file decoded as JSON Map.
+  Future<Map<String, dynamic>> getContentAsJsonMap([String version]) async {
```

Comment:
Can't we just return `Future<dynamic>` and call it `getJson([String version = versions.runtimeVersion])` ?